devito/dse/manipulation.py

```diff
@@ -284,14 +284,29 @@ def _topological_sort(exprs):
                 # Avoid cyclic dependences, such as
                 # Eq(f, f + 1)
                 continue
-            elif r.is_Indexed:
-                # Only scalars enforce an ordering
+            elif r.is_Indexed and e.lhs in retrieve_terminals(mapper[r].rhs):
+                # Cyclic dependencies stay "as is" ie
+                # [Eq(r1, u + 1), Eq(u, r1 + 2)]
                 continue
             else:
                 dag.add_edge(mapper[r], e, force_add=True)
     processed = dag.topological_sort()
 
     # Append tensor equations at the end in user-provided order
-    processed.extend(tensors)
+    processed_ordered = []
```

Comment:
OK after a very quick look... what I don't understand why is it not enough to add a few more edges to the DAG? 

Also, not sure if useful or not, but note that dag's `topological_sort` takes an optional argument `callback` to steer sorting

I think it's not neat that half of the algorithm is with DAG and half is post-processing. I'm fairly confident we can fix the bug with DAG+toposort.

As I said, I don't have much time these days, but I'll be happy to take a look on monday

Reply:
>I'm fairly confident we can fix the bug with DAG+toposort.

I'm not but feel free to try.